client/src/components/JurisdictionAdmin/BatchRoundDataEntry/BatchDetails.tsx

```diff
@@ -0,0 +1,472 @@
+    margin-bottom: 16px;
+    margin-top: 16px;
+    table-layout: fixed;
+    width: 100%;
+  }
+
+  &.${Classes.HTML_TABLE} tbody tr {
+    height: 56px;
+  }
+
+  &.${Classes.HTML_TABLE} td {
+    vertical-align: middle;
+  }
+`
+
+const BatchResultTallySheetButtonRow = styled(ButtonRow).attrs({
+  alignItems: 'center',
+  justifyContent: 'end',
+})`
+  min-height: 30px;
+`
+
+const VOTE_TOTALS_TAB_ID = 'vote-totals'
+
+interface IBatchResultTallySheetWithId extends IBatchResultTallySheet {
+  id: string
+}
+
+interface ITab {
+  id: string
+  name: string
+}
+
+function sheetWithId(
+  sheet: IBatchResultTallySheet
+): IBatchResultTallySheetWithId {
+  return { ...sheet, id: uuidv4() }
+}
+
+function sheetWithoutId(
+  sheet: IBatchResultTallySheetWithId
+): IBatchResultTallySheet {
+  return { name: sheet.name, results: sheet.results }
+}
+
+function defaultSheetName(sheetNumber: number) {
+  return `Sheet ${sheetNumber}`
+}
+
+function constructEmptySheet(
+  sheetName: string,
+  contest: IContest
+): IBatchResultTallySheet {
+  const results: { [choiceId: string]: number } = {}
+  contest.choices.forEach(choice => {
+    results[choice.id] = 0
+  })
+  return { name: sheetName, results }
+}
+
+function tabsFromSheets(sheets: IBatchResultTallySheetWithId[]): ITab[] {
+  if (sheets.length === 1) {
+    return [{ id: sheets[0].id, name: 'Vote Totals' }]
+  }
+  return [
+    { id: VOTE_TOTALS_TAB_ID, name: 'Vote Totals' },
+    ...sheets.map(sheet => ({ id: sheet.id, name: sheet.name })),
+  ]
+}
+
+interface IBatchDetailsProps {
+  areResultsFinalized: boolean
+  batch: IBatch
+  contest: IContest
+  saveBatchResults: (
+    resultTallySheets: IBatchResultTallySheet[]
+  ) => Promise<void>
+  setAreChangesUnsaved: (areChangesUnsaved: boolean) => void
+
+  // Require a key to ensure that the state within this component resets when a different batch is
+  // selected
+  key: string // eslint-disable-line react/no-unused-prop-types
+}
+
+const BatchDetails: React.FC<IBatchDetailsProps> = ({
+  areResultsFinalized,
+  batch,
+  contest,
+  saveBatchResults,
+  setAreChangesUnsaved,
+}) => {
+  const [sheets, setSheets] = useState<IBatchResultTallySheetWithId[]>(
+    (batch.resultTallySheets.length === 0
+      ? [constructEmptySheet(defaultSheetName(1), contest)]
+      : batch.resultTallySheets
+    ).map(sheetWithId)
+  )
+  const tabs = tabsFromSheets(sheets)
+  const [selectedTabId, setSelectedTabId] = useState(tabs[0].id)
+  const [isEditing, setIsEditing] = useState(false)
+
+  const currentSheetIndex = sheets.findIndex(
+    sheet => sheet.id === selectedTabId
+  )
+
+  // ---------- Handlers (start) ----------
+
+  const updateCurrentSheet = async (updatedSheet: IBatchResultTallySheet) => {
+    const updatedSheets = sheets.map((sheet, i) =>
+      i === currentSheetIndex ? { ...updatedSheet, id: sheet.id } : sheet
+    )
+
+    // Update client-side state first for immediate UI feedback
+    setSheets(updatedSheets)
```

Comment:
Yes good q - this comment got me thinking about a few things.

I wasn't seeing flickering but an expected lag in UI feedback (noticeable if you throttle your network speed in the Chrome dev tools) for certain actions, namely sheet additions and removals (the former is not relevant any more since we aren't saving new sheets until tallies are entered and saved).

I'm realizing now that the sheet update action doesn't need this optimistic client-side state update. We're displaying values from react-hook-form state (not the `sheets` state value) and disabling other UI elements until the API call completes. So the optimistic `setSheets` call doesn't actually do anything, and we're also providing clear user feedback.

The sheet removal action, however, does benefit from an optimistic client-side state update. If we wait for the API call to complete without this update, we'd want to introduce some kind of loading state to indicate that a sheet removal is in progress and block other UI actions in the interim. In this case, the immediate feedback of the sheet going away (while we continue processing the API update behind the scenes) feels like a better UX.

Reading the article that you linked did remind me of the need to revert optimistic state updates in the case of failure, hopefully a very rare case, but important for consistency / minimal user confusion.

Relevant commits: https://github.com/votingworks/arlo/pull/1661/commits/af7a4798c644c648f5c257a06862d259b325c62b, https://github.com/votingworks/arlo/pull/1661/commits/56ef3405a9ce6e3dc5c7419e515f629a97d03c47

While I'm sure I could push some of this logic into the React query layer, I'm gonna punt on that for now

Reply:
Updated ✅

Required a few new state values and switching the `Popover` from uncontrolled usage to controlled: https://github.com/votingworks/arlo/pull/1661/commits/ea96177f10e58b3a1b71b23ef47f77b2e7167adf

https://user-images.githubusercontent.com/12616928/199595160-2b30b72d-7c46-4900-8ab5-dcc284f20a4d.mov